### PR TITLE
Add `RetryWhen` operator.

### PR DESCRIPTION
**Pull request by danielt1263**
New pull request created as requested.

### CODE REVIEW

**Comment by kvaljeva** [comment]
@freak4pc are there any plans to merge this? I've noticed that there was an ongoing discussion where a separate PR was requested, but this ended up hanging about and being stale for a while.

Can we get this merged perhaps?

**Comment by freak4pc** [comment]
I have to be honest that I have almost no time to actively review PRs - especially more involved ones like this. I'd like to merge this but I need to find some time to get to it. Hopefully this week or the one after.

**Comment by danielt1263** [comment]
Is there something that someone else can do? I was hoping that since the operator is fully self contained now and wholly additive... and even comes with unit tests, it would be relatively easy to accept. Is there something I can do to help make it easier to review?

File: Sources/Operators/RetryWhen.swift [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        private let downstream: Downstream
+        private let errorSubject = PassthroughSubject<Upstream.Failure, Never>()
+        private var sink: Sink<Upstream, Downstream>?
+        private var cancellable: AnyCancellable?
+        
+        init(upstream: Upstream, downstream: Downstream, handler: @escaping (AnyPublisher<Upstream.Failure, Never>) -> Trigger) {
+            self.upstream = upstream
+            self.downstream = downstream
+            self.sink = Sink(
+				upstream: upstream,
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:58]: fix indentation

File: Sources/Operators/RetryWhen.swift [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+extension Publishers.RetryWhen {
+    class Subscription<Downstream>: Combine.Subscription where Downstream: Subscriber, Downstream.Input == Upstream.Output, Downstream.Failure == Upstream.Failure {
+        
+        private let upstream: Upstream
+        private let downstream: Downstream
+        private let errorSubject = PassthroughSubject<Upstream.Failure, Never>()
+        private var sink: Sink<Upstream, Downstream>?
+        private var cancellable: AnyCancellable?
+        
+        init(upstream: Upstream, downstream: Downstream, handler: @escaping (AnyPublisher<Upstream.Failure, Never>) -> Trigger) {
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:54]: split arguments into their own lines here
```suggestion
        init(
            upstream: Upstream,
            downstream: Downstream,
            handler: @escaping (AnyPublisher<Upstream.Failure, Never>) -> Trigger
        ) {
```

File: Sources/Operators/RetryWhen.swift [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        public func receive<S>(subscriber: S) where S: Subscriber, Failure == S.Failure, Output == S.Input {
+            subscriber.receive(subscription: Subscription(upstream: upstream, downstream: subscriber, handler: handler))
+        }
+    }
+}
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+extension Publishers.RetryWhen {
+    class Subscription<Downstream>: Combine.Subscription where Downstream: Subscriber, Downstream.Input == Upstream.Output, Downstream.Failure == Upstream.Failure {
+
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:47]: remove empty line
```suggestion
```

File: Sources/Operators/RetryWhen.swift [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    func retryWhen<Trigger>(_ notificationHandler: @escaping (AnyPublisher<Self.Failure, Never>) -> Trigger)
+        -> Publishers.RetryWhen<Self, Trigger, Output, Failure> where Trigger: Publisher {
+            .init(upstream: self, notificationHandler: notificationHandler)
+    }
+}
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public extension Publishers {
+    class RetryWhen<Upstream, Trigger, Output, Failure>: Publisher where Upstream: Publisher, Upstream.Output == Output, Upstream.Failure == Failure, Trigger: Publisher {
+
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:27]: remove empty line

File: Sources/Operators/RetryWhen.swift [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+//
+//  Created by Daniel Tartaglia on 3/21/20.
+//
+
+#if canImport(Combine)
+import Combine
+
+@available(OSX 10.15, iOS 13.0, tvOS 13.0, watchOS 6.0, *)
+public extension Publisher {
+
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:13]: remove empty line

File: Sources/Operators/RetryWhen.swift [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    /// Repeats the source publisher on error when the notifier emits a next value. If the source publisher errors and the notifier completes, it will complete the source sequence.
+    ///
+    /// - Parameter notificationHandler: A handler that is passed a publisher of errors raised by the source publisher and returns a publisher that either continues, completes or errors. This behavior is then applied to the source publisher.
+    /// - Returns: A publisher producing the elements of the given sequence repeatedly until it terminates successfully or is notified to error or complete.
+    func retryWhen<Trigger>(_ notificationHandler: @escaping (AnyPublisher<Self.Failure, Never>) -> Trigger)
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:18]: notificationHandler feels like a strange name, do you have other thoughts for naming here?

File: Sources/Operators/RetryWhen.swift [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+            self.cancellable = handler(errorSubject.eraseToAnyPublisher())
+                .sink(
+                    receiveCompletion: { [sink] completion in
+                        switch completion {
+                        case .finished:
+                            sink?.buffer.complete(completion: Subscribers.Completion<Downstream.Failure>.finished)
+                        case .failure(let error):
+                            if let error = error as? Downstream.Failure {
+                                sink?.buffer.complete(completion: Subscribers.Completion<Downstream.Failure>.failure(error))
+                            }
```
Comment (freak4pc) [inline Sources/Operators/RetryWhen.swift:75]:
```suggestion
                            sink?.buffer.complete(completion: >.finished)
                        case .failure(let error):
                            if let error = error as? Downstream.Failure {
                                sink?.buffer.complete(completion: .failure(error))
                            }
```

I think inference should work here?

File: Tests/RetryWhenTests.swift [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+	func testSuccessfulRetry() {
+		var times = 0
+
+		var expectedOutput: Int?
+
+		var completion: Subscribers.Completion<RetryWhenTests.MyError>?
+
+		subscription = Deferred(createPublisher: { () -> AnyPublisher<Int, MyError> in
+			defer { times += 1 }
```
Comment (freak4pc) [inline Tests/RetryWhenTests.swift:52]: can you switch to spaces? seems like everything is indented with Tabs here.

File: Tests/RetryWhenTests.swift [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+			nil
+		)
+		XCTAssertEqual(completion, .finished)
+	}
+
+	enum MyError: Swift.Error {
+		case someError
+		case retryError
+	}
+
```
Comment (freak4pc) [inline Tests/RetryWhenTests.swift:122]: remove empty line here please

File: Sources/Operators/RetryWhen.swift (hunk 6 above)
Comment (danielt1263) [reply Sources/Operators/RetryWhen.swift:18]: If I keep it as a noun phrase, I could see changing it into `errorStream`? As a verb phrase, maybe `bindError[Stream]`?

Comment (freak4pc) [reply Sources/Operators/RetryWhen.swift:18]: Can't it be just `trigger` ? By the way doesn't retryWhen usually accept the error itself? Or it always gets a publisher? I don't remember the RxSwifrt variant.

Comment (danielt1263) [reply Sources/Operators/RetryWhen.swift:18]: Sure, trigger is fine. The RxSwift variant receives an `Observable<Error>`.

Comment (danielt1263) [reply Sources/Operators/RetryWhen.swift:18]: Wait, I use `Trigger` as the output of the operator... Should I change that if the input is being named `trigger`?

Comment (freak4pc) [reply Sources/Operators/RetryWhen.swift:18]: No, I think it's fine :) You accept an external trigger and output your own trigger, I don't have an issue with it. You can name the generic RetryTrigger if you really want to.

Comment (danielt1263) [reply Sources/Operators/RetryWhen.swift:18]: The two `Trigger`s did cause a compile issue, so I named the external one `RetryTrigger` and the internal one `ErrorTrigger`...

**Review by freak4pc — APPROVED** [review]

**Pull request merged**